pkg/rc/memory_backed.go

```diff
@@ -0,0 +1,231 @@
+package rc
+
+import (
+	"strings"
+	"time"
+
+	"github.com/square/p2/pkg/kp"
+	"github.com/square/p2/pkg/labels"
+	"github.com/square/p2/pkg/logging"
+	"github.com/square/p2/pkg/pods"
+	"github.com/square/p2/pkg/util"
+)
+
+// I just copied this from kp.Store because it was a huge interface and I really only care about this one method!
+type kpStore interface {
+	SetPod(key string, manifest pods.Manifest) (time.Duration, error)
+}
+
+type memoryRc struct {
+	manifestSha string
+	logger      logging.Logger
+
+	kpStore        kpStore
+	nodeApplicator labels.Applicator
+	podApplicator  labels.Applicator
+
+	nodeSelector    labels.Selector
+	manifest        *pods.Manifest
+	replicasDesired int
+	disabled        bool
+}
+
+// memoryRc implements the ReplicationController interface.
+var _ ReplicationController = &memoryRc{}
+
+func NewMemoryBacked(
+	manifest *pods.Manifest,
+	nodeSelector labels.Selector,
+	kpStore kpStore,
+	nodeApplicator, podApplicator labels.Applicator,
+) (*memoryRc, error) {
+	mySha, err := manifest.SHA()
+	if err != nil {
+		return nil, err
+	}
+
+	rc := &memoryRc{
+		manifest:        manifest,
+		manifestSha:     mySha,
+		logger:          logging.DefaultLogger,
+		kpStore:         kpStore,
+		nodeApplicator:  nodeApplicator,
+		podApplicator:   podApplicator,
+		nodeSelector:    nodeSelector,
+		replicasDesired: 0,
+		disabled:        false,
+	}
+
+	return rc, nil
+}
+
+func (rc *memoryRc) ID() string {
+	return rc.manifestSha
+}
+
+func (rc *memoryRc) Manifest() *pods.Manifest {
+	return rc.manifest
+}
+
+func (rc *memoryRc) NodeSelector() labels.Selector {
+	return rc.nodeSelector
+}
+
+func (rc *memoryRc) IncreaseReplicas(n int) (int, error) {
+	// TODO should we error if increase would be > len(eligibleNodes)?
+	// Currently, that will cause an error in Meet()
+	rc.replicasDesired += n
+	return rc.replicasDesired, nil
+}
+
+func (rc *memoryRc) DecreaseReplicas(n int) (int, error) {
+	if n > rc.replicasDesired {
+		return rc.replicasDesired, util.Errorf("Cannot decrease desired replicas by %d from %d", n, rc.replicasDesired)
+	}
+	rc.replicasDesired -= n
+	return rc.replicasDesired, nil
+}
+
+func (rc *memoryRc) DesiredReplicas() int {
+	return rc.replicasDesired
+}
+
+func (rc *memoryRc) Disable() error {
+	rc.disabled = true
+	return nil
+}
+
+func (rc *memoryRc) Meet() error {
```

Comment:
This `Meet()` is analogous to the watch loop in the preparer - it seems desirable to want to recreate that same pattern of watching + responding.